README.md

## EM6819_Beacon Version 3.0.0

This package contains the Software Development Kit (SDK) for the following
EM Microelectronic products:

* EMBC01 (Bluetooth Low Energy Beacon)
* EMBC02 (Beacon with Accelerometer)
* EMBC03 (Beacon with Temperature Sensor)

### SDK Organization

This release is organized as follows:

    EM6819_Beacon             <-- Top level release directory
        doc
            HWDesign          <-- Device schematics and board layout
                CoinCell
                EMBC02
                EMBC03
                         
        include               <-- MCU dependent include files
            EM6819F6-A000
            EM6819FX-B300

        lib                   <-- ID Data and non-volatile mem (nvm) libraries

        project
            EMBC01            <-- 01 Coincell project located here
                build         <-- Build outputs (.hex, .map files)
            EMBC02            <-- 02 Accelerometer project located here
                build         <-- Build outputs (.hex, .map files)
            EMBC03            <-- 03 Temperature Sensor project located here
                build         <-- Build outputs (.hex, .map files)
            
        src                   <-- Primary source code for the project

        util                  <-- Beacon Programming and Update Tools

### Rebuilding

This release supports multiple products and a variety of on-board sensors.  Each configuration has its own Ride7 project file located in /project/EMBCxx, where 'xx' = 01, 02, or 03.  For example to build the EMBC02 hex file, open and build the /project/EMBC02/EMBC02.rprj project file.  The hex file will appear in the *build* subdirectory.

Each project pulls the source files it needs from a common pool in /src and /include.   

__NOTE:__ To compile this release it is necessary to update your RIDE7 installation with corrected libraries for the EM6819F6-B300 microcontroller used for the EMBCxx beacon design. Download the [*Ride7 Update for EM6819F6-B300*][2] from the distributions forum on emdeveloper.com. Unzip the contents and follow the instructions in the README file found there.

  [2]: https://forums.emdeveloper.com/index.php?threads/ride7-update-for-em6819f6-b300.57/

### IdData Support

As delivered, each hex file (01, 02, 03) includes IdData (iBeacon compatible) packets.  However, the iddata.a
library contains only stub functions. This means that if you recompile and program without any changes, your beacon will not transmit IdData packets.  To remedy this, you have several options.

* Obtain the fully-featured IdData library. 
    
>>The emBeacon ID source and object library is available free of charge from EM with proof of your agreement to [Apple's iBeacon License available at https://developer.apple.com/ibeacon][3].  You will receive an acknowledgment from Apple in pdf form.  Please send a copy of the first page, which contains your e-signature and contract number, to EM at [support@emdeveloper.com][4].  Be sure to state that you are requesting source code for the IdData library.
   [3]: https://developer.apple.com/ibeacon/
   [4]: mailto:support@emdeveloper.com


* Change ADVMODES_DEFAULT to select a state machine/packet selection that does not include IdData.  Values for this symbol are documented under the "beaconMode" entry in /util/ParameterDictionary.pdf in this SDK.

* Use the hex file as-is but program it with the scripts in the util directory and use a CSV file to update the beaconMode parameter.  This is a relatively simple process. If you have never done this before, see the PDF files in the util directory or review [this EMDeveloper blog post entitled "*Beacon Programming and Update Tools*"][1].

  [1]: https://www.emdeveloper.com/?p=455



RELEASENOTES.md

## Release Notes

### Version 3.0.0 Public Release

* Add full support for Eddystone packet types  
* Fix an EMBC02 bug whereby occasional +7g spikes were observed
* Misc BMA222E configuration changes for improved EMBC02 performance and power consumption 
* Accept _EM6819F6B300_  or _EM6819F6_B300_  as MCU definition. Fixes the build broken by a Ride7 IDE update.
* Fix #145, a bug whereby negative temperatures were reported incorrectly

### Version 2.8.0 Public Release

* Add support for the EMBC02 with BMA222E accelerometer.
* Add support for the EMBC03 with EM4325 (for temperature sensor).
* Add license header and EMUS copyright to all custom source. Add Source License pdf to distribution root.
* Fix #130, Program/Update scripts crash when log file is empty.
* Fix #133. Reduce delay between Power-On Self Test (POST) cycles from 50ms to 10ms and exclude AltBeacon advertisements from POST.
* Fix #134, prohibit locking the beacon in OFF state.
* Fix #135. Create separate projects for EMBC01, 02, and 03.
* Fix #136, Disable button pull-up during current test.  Also put 9301 in sleep mode instead of idle.
* Fix #140. Use correct flags per iBeacon spec.


* Tiny/Low cost support is dropped beginning with release 2.8.0.  References to Tiny/Low cost still exist in the code but there is no project file and the code does not build in Tiny/Low cost configuration.      
 
### Version 2.7 Public Release

This release supports the COiN product. Tiny and LowCost support remains in
this release, but it was not tested.

* Rotates beacon type each wake-up cycle to provide separation between advertisement packets as required by the Bluetooth standard.
* Implements AltBeacon packets for planned production of COiN devices for CES.

__Known issues__

* The Tiny resets when battery level drop below ~1.3V
* The program_beacon and update_beacon script displays the deviceAddress values
without leading "0x" on the console and in the log file.


### Version 2.6 Public Release

* Fixes #124 Always advertise 9301 autocal result during POST.
* Fixes #125 Add production test for 6819/9301 sleep current
* Fixes #118, Tiny/LC battery voltage is incorrect
* Fixes #121, shortening Parameter Dictionary by ~ 50%.
* Fixes #38 (re-opened) long button-press at wake-up behavior
* Fixes #89 Clean-up TODO issues in firmware (opened #119 as a result)
* Fixes #117 Temperature Offset default now zero
* Fixes #113 Tiny Beacon Packet Data Errors (opened #118 as separate matter)
* Fixes #109, Create TemperatureOffset Parameter
* Fixes #110, Use OpenSensor Parameter to Specify Both Sensor AND Event Counter
* Fix MISO setup.  A small percentage of coincells did not advertise during POST prior to fix.
* Fix an error affecting any platform with di07.


### Version 2.5 Public Release

Support for COiN/Tiny/LowCost with the same code base, same feature set.
Programming utilities integrated into the SDK.
Programming utilities UI simplified.

* program_beacon [options] <firmware.hex> <parameters.csv>
* update_beacon [options] <firmware.hex> <parameters.csv>
* update_beacon [options] <firmware.hex>
* update_beacon [options] <parameters.csv>
* For a complete description, see ../util/usage.md or usage.pdf

COiN Schematic and mechanical drawings updated

__Numerous minor fixes and enhancements:__

* Add FET control for EM9301 power (#75)
* Go into a permanent (until battery is replaced) low power state if the 9301 passes self test but
    fails calibration. (#99)
* Add the Proximity beacon state machine.  (#97)
* Bug fix - update the power field in id data packet dynamically  (#103)
* Battery linearization  (#84)
* Remove temperature from default advertisement.  Replace with modelID.  Temperature is optionally
    selectable as open sensor field number 4.  (#87)
* Apply random 0..10ms jitter to advertising interval  (#82)
* Sample temperature, battery, and lux (if present) every 4th advertising cycle
* Change the beacon name to use a 5-digit value derived from minor ID  (#79)
* Support for energy harvesting apps.  Skip POST. Other enhancements when button is pressed at powerup.  (#38)
* Bug fix - eliminate extra button count at startup  (#9)
* Improved button debounce  (#24)
* The default state machine for Tiny/LC is simple on/off.
* Whenever the platform has a light sensor, the default openSensor field is the lux value.


### Version 2.4 Public Release

Changes in this public release (includes all work listed in previous releases):

* Enhancement #78:  Remove Customization Scripts From Project
* Enhancement #77:  Set default UUID to 699EBC80-E1F3-11E3-9A0F-0CF3EE3BC012
* Enhancement #69:  Eliminate HCI "disable RF periodic autocal" when only sensor OR id data is transmitted
* Enhancement #68:  Improve Temp Measurement Stability
* Bug fix #67:  Measure Battery Before self-test
* Bug fix #66:  Revise RF Cal Self-Test Limits
* Bug fix #65: Long press only works correctly when beacon is on.
* Bug fix #64: Error in 6819 distribution file irqmsk.h - ClrPending_IRQ_SlpCnt() macro.
* Enhancement #61: Temperature-Compensate the 8KHhz RC Clock
* Documentation Fix #58:  Release Notes Incorrectly Define GASP_SCK Sense
* Enhancement #57: Default Parameters (related to #39)
* Enhancement #54: Min Advertising Interval 100ms
* Enhancement #50: Do not beacon below 2.5V (EM6818) or 0.9V (EM6819)
* Enhancement #40: Obtain Device Parameters from Row 62.  Default device address and
advertising parameters are written to write-once memory at the factory.
* Enhancement #39: Multiple State Machine, Chosen by Parameterization.  The customization
script can specifiy the advertisement modes.  1) sensor data only 2) iddata only 3) both
4) coin standard state machine (default)
* Enhancement #35: New Button Behavior.  Coin standard and Coin custom state machines
implemented per published Fact Sheets.
* Enhancement #31: Beacon Interval set on 100ms increments.  The customization script requires
the beacon interval to be specificed on 100ms boundaries, with 100ms minimum.

src/hci.c
/**
 ** ############################################################################
 **
 ** @file  hci.c
 ** @brief Common HCI interface functions
 **
 **        Processor - EM6819
 **        Radio     - EM9301
 **
 ** DESCRIPTION:
 **    This module has funtions that performs the HCI interface between
 **    the EM6819 processor and the EM9301 radio.
 **
 ** Copyright (c) 2015 EM Microelectronic-US Inc. All rights reserved.
 ** Developed by Glacier River Design, LLC 
 **
 ** ############################################################################
 ** EM Microelectronic-US Inc. License Agreement
 ** 
 ** Please read this License Agreement ("Agreement") carefully before 
 ** accessing, copying, using, incorporating, modifying or in any way providing 
 ** ("Using" or "Use") this source code.  By Using this source code, you: (i) 
 ** warrant and represent that you have obtained all authorizations and other 
 ** applicable consents required empowering you to enter into and (ii) agree to 
 ** be bound by the terms of this Agreement.  If you do not agree to this 
 ** Agreement, then you are not permitted to Use this source code, in whole or 
 ** in part.
 ** 
 ** Pursuant to the terms in the accompanying software license agreement and 
 ** Terms of Use located at: www.emdeveloper.com/emassets/emus_termsofuse.html 
 ** (the terms of each are incorporated herein by this reference) and subject to 
 ** the disclaimer and limitation of liability set forth below, EM  
 ** Microelectronic US Inc. ("EM"), grants strictly to you, without the right to 
 ** sublicense, a non-exclusive, non-transferable, revocable, worldwide license 
 ** to use the source code to modify the software program for the sole purpose 
 ** of developing object and executable versions that execute solely and 
 ** exclusively on devices manufactured by or for EM or your products that use 
 ** or incorporate devices manufactured by or for EM; provided that, you clearly 
 ** notify third parties regarding the source of such modifications or Use.
 ** 
 ** Without limiting any of the foregoing, the name "EM Microelectronic-US 
 ** Inc." or that of any of the EM Parties (as such term is defined below) must 
 ** not be Used to endorse or promote products derived from the source code 
 ** without prior written permission from an authorized representative of EM 
 ** Microelectronic US Inc.
 ** 
 ** THIS SOURCE CODE IS PROVIDED "AS IS" AND "WITH ALL FAULTS", WITHOUT ANY 
 ** SUPPORT OR ANY EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED 
 ** TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 ** PURPOSE ARE DISCLAIMED.  ALSO, THERE IS NO WARRANTY OF NON-INFRINGEMENT, 
 ** TITLE OR QUIET ENJOYMENT.
 ** 
 ** IN NO EVENT SHALL EM MICROELECTRONIC US INC., ITS AFFILIATES, PARENT AND 
 ** ITS/THEIR RESPECTIVE LICENSORS, THIRD PARTY PROVIDERS, REPRESENTATIVES, 
 ** AGENTS AND ASSIGNS ("COLLECTIVLEY, "EM PARTIES") BE LIABLE FOR ANY DIRECT, 
 ** INDIRECT, INCIDENTAL, SPECIAL, PUNITIVE, EXEMPLARY, OR CONSEQUENTIAL 
 ** DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
 ** SERVICES; LOSS OF USE, DATA, EQUIPMENT, SYSTEMS, SOFTWARE, TECHNOLOGY, 
 ** SERVICES, GOODS, CONTENT, MATERIALS OR PROFITS; BUSINESS INTERRUPTION OR 
 ** OTHER ECONOMIC LOSS OR ANY CLAIMS BY THIRD PARTIES (INCLUDING BUT NOT 
 ** LIMITED TO ANY DEFENSE THEREOF) HOWEVER CAUSED AND ON ANY THEORY OF 
 ** LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
 ** NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOURCE 
 ** CODE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.  NOTWITHSTANDING 
 ** ANYTHING ELSE TO THE CONTRARY, IN NO EVENT WILL THE EM PARTIES' AGGREGATE 
 ** LIABILITY UNDER THIS AGREEMENT OR ARISING OUT OF YOUR USE OF THE SOURCE 
 ** CODE EXCEED ONE HUNDRED U.S. DOLLARS (U.S. $100).
 ** 
 ** Please refer to the accompanying software license agreement and Terms of 
 ** Use located at: www.emdeveloper.com/emassets/emus_termsofuse.html to better 
 ** understand all of your rights and obligations hereunder. 
 ** ############################################################################
 */


#include "Types.h"
#include "Macro.h"
#include "Periph.h"

#include "ports.h"
#include "irqmsk.h"
#include "spi.h"
#include "config.h"
#include "params.h"
#include "hci.h"
#include "hci_9301.h"
#include "spi_6819.h"

/**
 * @brief Advertise Enable command packet
 * Sent by the @ref HCI_LE_Set_Advertise_Enable and
 * @ref SendAdvertisement functions
 */
static HCI_LE_Set_Advertise_Enable_Packet_t HCI_LE_Set_Advertise_Enable_pkt=
{
   .hdr.length=1,
   .hdr.opcode=HCI_LE_SET_ADVERTISE_ENABLE_OPCODE,
   .data=0,
};

/**
 * @brief buffer for data received via spi
 * @brief number of valid bytes in spi_inbuf
 */
UINT8 spi_inbuf[SPI_INBUF_SIZE];
UINT8 spi_inBytes = 0;

/**
*******************************************************************************
* @brief Waits for an 9301 HCI command to complete
*
* Waits on IRQ signal from EM9301 to go high. Then reads data, if any from SPI.
*
* @param  none
* @return nothing
*******************************************************************************/
void HCI_WaitCommandComplete(void)
{
   UINT16 loopTimeout = 5000;

   spi_inBytes = 0;
   do
   {
      if(RegPADIn & BIT2)
      {
         spi_inBytes = HCI_LE_Read(spi_inbuf, SPI_INBUF_SIZE);
         break;
      }
   } while(loopTimeout--);
}

/**
******************************************************************************
* @brief Sends a HCI Command to the EM9301
*
* Reads the HCI Command size from the HCI_Packet_hdr_t
* @n   Starts the HCI SPI interface
* @n   Sends the data
* @n   Stops the SPI interface
* @n   Waits for a reply
* @n returns the status from the reply packet
*
* @param pData    pointer to an HCI command,
*                  the length is embedded in the HCI command
* @return         status read from the HCI interface
******************************************************************************/
void HCI_LE_Send(UINT8* pData)
{
   UINT8 len;
   struct pkt
   {
     HCI_Packet_hdr_t hdr;
   } *p;

   //get the length and launch the spi transaction
   p = (struct pkt *)pData;
   len = p->hdr.length+sizeof(HCI_Packet_hdr_t);
   SPI_Start_Send();

   UINT8 i;
   for(i = 0; i < 1; i++);    // necessary for gcc 3.2 to generate good code (?!!)
   switch(len)
   {
      // Deliberate coding standard violation: no 'break' statements, and no default.
      // This is an unrolled loop with multiple entry points. Not all are needed but
      // its more efficient to let it compute the branch address (a computed goto)
      // then to do test and branch for the 4 or 5 cases needed.
      //
      // Note, the spi bit rate and the 6819 instruction rate are essentially equal
      // so efficient transmit code is important.
      // The instruction costs are
      //                            ___Time____   ________Space________
      //     this switch statement: 15 +  2*len    85 = 15 + 2*max(len) GCC 3.2
      //      unless 3.2 screws up: 15 +  8*len   298 = 15 + 8*max(len) GCC 3.2
      //      GCC 4.4               12 +  2*len    82 = 12 + 2*max(len) GCC 4.4
      //     a do...while loop:     12 +  6*len    18
      //     a 'for' loop:           4 + 11*len    15
      //
      // The shortest packet is len = 4 so the switch cuts the overhead in half
      // and way by nearly 5X for a 35 byte packet.

      case 35:  SPI_Putc(*pData++);
      case 34:  SPI_Putc(*pData++);
      case 33:  SPI_Putc(*pData++);
      case 32:  SPI_Putc(*pData++);
      case 31:  SPI_Putc(*pData++);
      case 30:  SPI_Putc(*pData++);
      case 29:  SPI_Putc(*pData++);
      case 28:  SPI_Putc(*pData++);
      case 27:  SPI_Putc(*pData++);
      case 26:  SPI_Putc(*pData++);
      case 25:  SPI_Putc(*pData++);
      case 24:  SPI_Putc(*pData++);
      case 23:  SPI_Putc(*pData++);
      case 22:  SPI_Putc(*pData++);
      case 21:  SPI_Putc(*pData++);
      case 20:  SPI_Putc(*pData++);
      case 19:  SPI_Putc(*pData++);
      case 18:  SPI_Putc(*pData++);
      case 17:  SPI_Putc(*pData++);
      case 16:  SPI_Putc(*pData++);
      case 15:  SPI_Putc(*pData++);
      case 14:  SPI_Putc(*pData++);
      case 13:  SPI_Putc(*pData++);
      case 12:  SPI_Putc(*pData++);
      case 11:  SPI_Putc(*pData++);
      case 10:  SPI_Putc(*pData++);
      case  9:  SPI_Putc(*pData++);
      case  8:  SPI_Putc(*pData++);
      case  7:  SPI_Putc(*pData++);
      case  6:  SPI_Putc(*pData++);
      case  5:  SPI_Putc(*pData++);
      case  4:  SPI_Putc(*pData++);
      case  3:  SPI_Putc(*pData++);
      case  2:  SPI_Putc(*pData++);
      case  1:  SPI_Putc(*pData);
   }
   SPI_Stop_Send();
}

/**
******************************************************************************
* @brief Reads data from the HCI interface
*
* Starts the HCI SPI interface
* @n Reads the data
* @n Stops the SPI interface
*
* @param pData Pointer to a data buffer
* @param len   Size of the data buffer
* @return      Number of bytes read
******************************************************************************/
UINT8 HCI_LE_Read(UINT8* pData,UINT8 len)
{
   UINT8 nbytes;

   // Determine if any data is available
   if(SPI_Start_Read() == 0)
   {
      //no data
      return 0;
   }

   //read each data byte from the radio
   for(nbytes=0; nbytes<len; nbytes++)
   {
       if(SPI_Getc(&pData[nbytes]) == 0)
       {
           break;
       }
   }

   //end the read process
   SPI_Stop_Read();

   //return the number of bytes
   return nbytes;
}

/**
******************************************************************************
* @brief sends the Advertise Enable command to the HCI interface
*
* @param  enable 1 for start, 0 for stop
* @return nothing
******************************************************************************/
void HCI_LE_Set_Advertise_Enable(UINT8 enable)
{
   HCI_LE_Set_Advertise_Enable_pkt.data = enable;
   HCI_LE_Send((UINT8*)&HCI_LE_Set_Advertise_Enable_pkt);
   HCI_WaitCommandComplete();
}

#define HCI_LE_SetAdvertiseStartStop(flag) \
   HCI_LE_Set_Advertise_Enable_pkt.data = flag; \
   HCI_LE_Send((UINT8*)&HCI_LE_Set_Advertise_Enable_pkt)

#define HCI_LE_Start_Advertising() HCI_LE_SetAdvertiseStartStop(HCI_LE_Set_Advertise_Start)
#define HCI_LE_Stop_Advertising()  HCI_LE_SetAdvertiseStartStop(HCI_LE_Set_Advertise_Stop)

/**
******************************************************************************
* @brief sends an advertisement
*
* Sends a HCI_LE_Set_Advertise_Enable command to start advertising
* @n waits for the transmission
* @n Sends a HCI_LE_Set_Advertise_Enable command to stop advertising
* @note should have a parameter for wait time?
*
* @param  none
* @return nothing
******************************************************************************/
void SendAdvertisement(void)
{
   UINT8 saveBtnIntStat;

   // -- Send HCI command to set Adv data --
   Save_BTN_IRQ_Status(saveBtnIntStat);
   HCI_LE_Start_Advertising();
   HCI_WaitCommandComplete();

   // wait until all three advertisements (one per advertising channel) are sent
   #if CAPABILITY_EM9301_SIGNALS_RF_ACTIVITY
   if(RFActSigIsSupported())
   {
      while (!RF_ActivityPresent()) {}  // wait till advertisement starts
      while ( RF_ActivityPresent()) {}  // then wait till it stops
   }
   else
   #endif
   {
      WaitNmSec(20);
   }
   Restore_BTN_IRQ_Status(saveBtnIntStat);

   //end the advertise messaging
   HCI_LE_Stop_Advertising();
   HCI_WaitCommandComplete();
};

//--------------- END OF FILE ------------------------//
